src/Dofs/DofHandler2.jl

```diff
@@ -0,0 +1,245 @@
+    end
+end
+
+ndofs(dh::DofHandler) = maximum(dh.cell_dofs) # TODO: This is not very nice
+ndofs_per_cell(dh::DofHandler, cell::Int=1) = dh.cell_dofs_offset[cell+1] - dh.cell_dofs_offset[cell] # TODO: This is not very nice
+isclosed(dh::DofHandler) = dh.closed[]
+nfields(dh::DofHandler) = length(dh.field_names)
+
+function Base.push!(dh::DofHandler, name::Symbol, ip::Interpolation, dim::Int)
+    @assert !isclosed(dh)
+    @assert !in(name, dh.field_names)
+    push!(dh.field_names, name)
+    push!(dh.field_interpolations, ip)
+    push!(dh.field_dims, dim)
+    nothing
+end
+
+# sort and return true (was already sorted) or false (if we had to sort)
+function sortedge(edge::Tuple{Int,Int})
+    a, b = edge
+    a < b ? (return (edge, true)) : (return ((b, a), false))
+end
+
+sortface(face::Tuple{Int,Int}) = minmax(face[1], face[2])
+function sortface(face::Tuple{Int,Int,Int})
+    a, b, c = face
+    b, c = minmax(b, c)
+    a, c = minmax(a, c)
+    a, b = minmax(a, b)
+    return (a, b, c)
+end
+
+# close the DofHandler and distribute all the dofs
+function close!(dh::DofHandler{dim}) where {dim}
+    @assert !isclosed(dh)
+
+    # `vertexdict` keeps track of the visited vertices. We store the global vertex
+    # number and the first dof we added to that vertex.
+    vertexdicts = [Dict{Int,Int}() for _ in 1:nfields(dh)]
+
+    # `edgedict` keeps track of the visited edges, this will only be used for a 3D problem
+    # An edge is determined from two vertices, but we also need to store the direction
+    # of the first edge we encounter and add dofs too. When we encounter the same edge
+    # the next time we check if the direction is the same, otherwise we reuse the dofs
+    # in the reverse order
+    edgedicts = [Dict{Tuple{Int,Int},Tuple{Int,Bool}}() for _ in 1:nfields(dh)]
+
+    # `facedict` keeps track of the visited faces. We only need to store the first dof we
+    # added to the face; if we encounter the same face again we *always* reverse the order
+    # In 2D a face (i.e. a line) is uniquely determined by 2 vertices, and in 3D a
+    # face (i.e. a surface) is uniquely determined by 3 vertices.
+    facedicts = [Dict{NTuple{dim,Int},Int}() for _ in 1:nfields(dh)]
+
+    # celldofs are never shared between different cells so there is no need
+    # for a `celldict` to keep track of which cells we have added dofs too.
+
+    # We create the `dofinfo` matrix with precomputed information for each
+    # interpolation since that allows for looping over the cells first,
+    # and the interpolations inside without using a function barrier
+    #
+    # dofinfo =
+    # | nvertexdofs(field#1) | nvertexdofs(field#2) | ...
+    # | nedgedofs(field#1)   | nedgedofs(field#2)   | ...
+    # | nfacedofs(field#1)   | nfacedofs(field#2)   | ...
+    # | ncelldofs(field#1)   | ncelldofs(field#2)   | ...
+
+    dofinfo = fill(0, 4, nfields(dh))
+    for (i, interpolation) in enumerate(dh.field_interpolations)
```

Comment:
`interpolation` here is the only non-inferred variable that shows up for a `@code_warntype close!(dh)` call. Which should be OK (?).

Reply:
Yes, that should be fine.